# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/workers/WorkerRunnable.h
# Commit: 9e3ed11f4e8a
# Full Hash: 9e3ed11f4e8a3e4af5fd56d084f142c50eb30bb8
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2024-05-02 16:06:59
# Description:
#   Bug 1892644 - Handling WorkerThreadRunnable::Run() after Worker is "Dead". r=asuth
#   
#   In bug 1769913, we remove the WorkerThreadRunnable's raw pointer to the corresponding WorkerPrivate and expect the corresponding WorkerPrivate to be obtained by GetCurrentThreadWorkerPrivate() when WorkerThreadRunnable::Run executing.
#   
#   In general, the assumption is correct. However, it could be violated in the following two situations.
# ==============================================================================

diff -r 810c65d8b2d3 -r 9e3ed11f4e8a dom/workers/WorkerRunnable.h
--- a/dom/workers/WorkerRunnable.h	Thu May 02 06:34:16 2024 +0000
+++ b/dom/workers/WorkerRunnable.h	Thu May 02 06:37:51 2024 +0000
@@ -13,8 +13,10 @@
 #include "mozilla/Atomics.h"
 #include "mozilla/RefPtr.h"
 #include "mozilla/ThreadSafeWeakPtr.h"
+#include "mozilla/dom/WorkerPrivate.h"
 #include "mozilla/dom/WorkerRef.h"
 #include "mozilla/dom/WorkerStatus.h"
+#include "mozilla/dom/quota/CheckedUnsafePtr.h"
 #include "nsCOMPtr.h"
 #include "nsIRunnable.h"
 #include "nsISupports.h"
@@ -33,8 +35,6 @@
 namespace dom {
 
 class Worker;
-class WorkerParentRef;
-class WorkerPrivate;
 
 class WorkerRunnable : public nsIRunnable
 #ifdef MOZ_COLLECTING_RUNNABLE_TELEMETRY
@@ -225,6 +225,8 @@
 };
 
 class WorkerThreadRunnable : public WorkerRunnable {
+  friend class WorkerPrivate;
+
  public:
   NS_INLINE_DECL_REFCOUNTING_INHERITED(WorkerThreadRunnable, WorkerRunnable)
 
@@ -261,6 +263,18 @@
   // method. Avoids infinite recursion when a subclass calls Run() from inside
   // Cancel(). Only checked and modified on the target thread.
   bool mCallingCancelWithinRun;
+
+  // If dispatching a WorkerThreadRunnable before Worker initialization complete
+  // in worker thread, which are in WorkerPrivate::mPreStartRunnables, when
+  // GetCurrentThreadWorkerPrivate() might get an invalid WorkerPrivate for
+  // WorkerThreadRunnable::Run() because it is in Worker's shutdown.
+  //
+  // This is specific for cleanup these pre-start runnables if the shutdown
+  // starts before Worker executes its event loop.
+  // This member is only set when the runnable is dispatched to
+  // WorkerPrivate::mPreStartRunnables. Any other cases to use this
+  // WorkerPrivate is always wrong.
+  CheckedUnsafePtr<WorkerPrivate> mWorkerPrivateForPreStartCleaning;
 };
 
 // This runnable is used to send a message to a worker debugger.